Sources/SwiftProtobufPluginLibrary/BytecodeWriter.swift

```diff
@@ -0,0 +1,105 @@
+// Copyright (c) 2014 - 2025 Apple Inc. and the project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See LICENSE.txt for license information:
+// https://github.com/apple/swift-protobuf/blob/main/LICENSE.txt
+//
+// -----------------------------------------------------------------------------
+
+import Foundation
+
+/// Writes bytecode into a string that is suitable for use as a Swift string literal (i.e., for
+/// appending to generated code).
+///
+/// See `SwiftProtobuf.BytecodeInterpreter` for more information on the format used here.
+public struct BytecodeWriter<Instruction: RawRepresentable> where Instruction.RawValue == UInt64 {
+    /// The Swift string literal that represents the written bytecode, including the delimiting
+    /// quotes.
+    public var stringLiteral: String { #""\#(code)""# }
+
+    /// The contents of the Swift string literal representing the written bytecode, without the
+    /// delimiting quotes.
+    private var code: String = ""
+
+    /// Creates a bytecode writer that uses the given program format.
+    ///
+    /// - Parameter programFormat: The program format of the bytecode stream to write. Currently,
+    ///   the only valid value is zero.
+    public init(programFormat: Int = 0) {
+        if programFormat != 0 {
+            fatalError("Unexpected bytecode program format \(programFormat)")
+        }
+        writeUInt64(UInt64(programFormat))
+    }
+
+    /// Writes the integer opcode corresponding to the given instruction to the bytecode stream.
+    public mutating func writeOpcode(of instruction: Instruction) {
+        writeUInt64(instruction.rawValue)
+    }
+
+    /// Writes a signed 32-bit integer to the bytecode stream.
+    ///
+    /// This is provided as its own primitive operation because 32-bit values are extremely common
+    /// as field numbers (0 to 2^29-1) and enum cases (-2^31 to 2^31-1). In particular for enum
+    /// cases, using this function specifically for those cases avoids making mistakes involving
+    /// sign- vs. zero-extension between differently-sized integers.
+    public mutating func writeInt32(_ value: Int32) {
+        // `Int32`s are stored by converting them bit-wise to a `UInt32` and then zero-extended to
+        // `UInt64`, since this representation is smaller than sign-extending them to 64 bits.
+        writeUInt64(UInt64(UInt32(bitPattern: value)))
+    }
+
+    /// Writes an unsigned 64-bit integer to the bytecode stream.
+    public mutating func writeUInt64(_ value: UInt64) {
+        func append(_ value: UInt64) {
+            // Print the normal scalar if it's ASCII-printable so that we only use longer `\u{...}`
+            // sequences for those that are not.
+            if isprint(Int32(truncatingIfNeeded: value)) != 0 {
+                self.append(escapingIfNecessary: UnicodeScalar(UInt32(truncatingIfNeeded: value))!)
+            } else {
+                code.append(String(format: "\\u{%x}", value))
```

Comment:
Is it worth picking off `\0` as the one special case?  (tab, newline, etc don't seem worth it, but if `\0` is common, it would slightly shorten strings and there's no code on the decoding side.

Reply:
Yeah, added `\0` as a special case now.